### ISSUE TITLE
"too large document" error when looking up state for a call

### ISSUE THREAD

**Issue by dangelog** [apitrace/apitrace#323]
If you have some big textures (4096x4096 or so) active when looking up state, the lookup will fail, and the UI will report in the status bar "too large document".

That's actually an error coming from Qt -- QJsonDocument is unable to parse the massive JSON string containing the texture. Apparently there's some hardcoded limit (2^27) on the length of a string and we end up over that limit.

http://code.woboq.org/qt5/qtbase/src/corelib/json/qjsonparser.cpp.html#593
http://code.woboq.org/qt5/qtbase/src/corelib/json/qjson_p.h.html#QJsonPrivate::Value::::MaxSize

I've no robust ideas about how to work around this...
* switch from JSON to another (custom) format?
* zlib-compress the image payload in all cases not currently handled by the PNG compression?
* stop using Qt's JSON classes?

**Comment by dangelog** [comment]
beginMember("__data__");
    std::stringstream ss;

    if (image->channelType == image::TYPE_UNORM8) {
        image->writePNG(ss);
    } else {
        image->writePNM(ss);
    }

    const std::string & s = ss.str();
    writeBase64(s.data(), s.size());
    endMember(); // __data__

Is this writing the image into a PNM and then base64 encoding the result into JSON? That's just asking for 3 times the size (PNM) * 4/3 (Base64) = 4 times time the memory space... Which also explains why 4k hits the problem: 4096*4096*4 byte = 64MB, *4 for the representation => 256MB > 2^27 limit

**Comment by dangelog** [comment]
Quick'n'dirty horrible workaround :)

```
diff --git a/gui/apisurface.cpp b/gui/apisurface.cpp
index 22de5f0..ad1a3bb 100644
--- a/gui/apisurface.cpp
+++ b/gui/apisurface.cpp
@@ -34,6 +34,8 @@ struct ByteArrayBuf : public std::streambuf
 void ApiSurface::contentsFromBase64(const QByteArray &base64)
 {
     m_base64Data = base64;
+    if (base64.isEmpty())
+        return;
 
     /*
      * We need to do the conversion to create the thumbnail
diff --git a/retrace/json.cpp b/retrace/json.cpp
index 597b62f..20d940e 100644
--- a/retrace/json.cpp
+++ b/retrace/json.cpp
@@ -335,7 +335,9 @@ JSONWriter::writeImage(image::Image *image,
         image->writePNM(ss);
     }
 
-    const std::string & s = ss.str();
+    std::string s = ss.str();
+    if (s.size() > (1 << 24))
+       s.clear();
     writeBase64(s.data(), s.size());
     endMember(); // __data__
 
```

**Comment by jrfonseca** [comment]
> Is this writing the image into a PNM and then base64 encoding the result into JSON? 

Yep.  But it should increase by 30%, not 3 times.

I don't know if there's a more efficient way of encoding binary data to JSON.

Another possibility would be to use UTF-8 strings, but the length of the escaped strings would be bigger in average.  See http://stackoverflow.com/questions/1443158/binary-data-in-json-string-something-better-than-base64

The ideal would be to abandon JSON for retrace->qapitrace communication, and use some binary format.  Either some "binary JSON" variant, or the apitrace trace format itself.  All binary json formats have pros/cons.  The apitrace trace format is not currently adequate for this -- it's missing a dictionary/map -- and it can't be easily parsed from scripting languages.

Overall, I suspect the ideal would be indeed to use apitrace trace format for this.

**Comment by jrfonseca** [comment]
I pushed a comprehensive fix for this to https://github.com/apitrace/apitrace/tree/ubjson

I ended up changing my mind -- I didn't reuse the .trace format for the GUI communication.  The .trace format was not adequate and needs to stay backward compatible, while the GUI can change any time we want, so I ended up using UBJSON that VOGL also uses.  Even though it is JSON compatible there's a way of encoding blobs efficiently.
